datalad/support/sshconnector.py

```diff
@@ -0,0 +1,165 @@
+        self.host = host
+        self.ctrl_path = ctrl_path
+        self.cmd_prefix = ["ssh", "-S", self.ctrl_path, self.host]
+
+    def __call__(self, cmd):
+        """
+        Parameters
+        ----------
+        cmd: list or str
+          command to run on the remote
+
+        Returns
+        -------
+        tuple
+          stdout, stderr
+        """
+
+        # TODO: Do we need to check for the connection to be open or just rely
+        # on possible ssh failing?
+
+        ssh_cmd = self.cmd_prefix + cmd if isinstance(cmd, list) \
+            else sh_split(cmd, posix=not on_windows)
+            # windows check currently not needed, but keep it as a reminder
+
+        # TODO: pass expect parameters from above?
+        # Hard to explain to toplevel users ... So for now, just set True
+        return self.runner.run(ssh_cmd, expect_fail=True, expect_stderr=True)
+
+    @property
+    def runner(self):
+        if self._runner is None:
+            self._runner = Runner()
+        return self._runner
+
+    def open(self):
+        # start control master:
+        cmd = "ssh -o ControlMaster=auto -o \"ControlPath=%s\" " \
+              "-o ControlPersist=yes %s exit" % (self.ctrl_path, self.host)
+        lgr.debug("Try starting control master by calling:\n%s" % cmd)
+        proc = Popen(cmd, shell=True)
+        proc.communicate(input="\n")  # why the f.. this is necessary?
+
+    # TODO: Probably not needed as an explicit call.
+    # Destructor should be sufficient.
+    def close(self):
+        # stop controlmaster:
+        cmd = ["ssh", "-O", "stop", "-S", self.ctrl_path, self.host]
+        try:
+            self.runner.run(cmd, expect_stderr=True, expect_fail=True)
+        except CommandError as e:
+            if "No such file or directory" in e.stderr:
+                # nothing to clean up
+                pass
+            else:
+                raise
+
+
+@auto_repr
+class SSHManager(object):
+    """Keeps ssh connections to share. Serves singleton representation
+    per connection.
+    """
+
+    def __init__(self):
+        not_supported_on_windows("TODO: Make this an abstraction to "
+                                 "interface platform dependent SSH")
+
+        self._connections = dict()
+
+        self.socket_dir = AppDirs('datalad', 'datalad.org').user_config_dir
+        assure_dir(self.socket_dir)
+
+    def get_connection(self, url):
+        """Get a singleton, representing a shared ssh connection to `url`
+
+        Parameters
+        ----------
+        url: str
+          ssh url
+
+        Returns
+        -------
+        SSHConnection
+        """
+
+        # parse url:
+        parsed_target = urlparse(url)
+
+        # Note: The following is due to urlparse, not ssh itself!
+        # We probably should find a nice way to deal with anything,
+        # ssh can handle.
+        if parsed_target.scheme != 'ssh':
+            raise ValueError("Not an SSH URL: %s" % url)
```

Comment:
relates to my note above if decide to assume that any "[user@]host:[path]" is for the ssh connection